src/core/rate_limiter.h

```diff
@@ -0,0 +1,263 @@
+
+  /// Creates a rate limiter object which will funnel the requests to
+  /// the model instances. A typical lifetime of the model instance within
+  /// RateLimiter transition from available -> staged -> allocated -> available.
+  /// The transition from available to staged occurs when a request is
+  /// registered for the model. Depending upon the resource availabilty and
+  /// priority, the RateLimiter will transition an instance to allocated state
+  /// at some point in the future. The staged state is skipped when
+  /// configured to ignore the resource constraints. The cycle in this case
+  /// will be available -> allocated -> available.
+  /// \param ignore_resources_and_priority Whether or not to ignore resource
+  /// constraints and cross-model priority. An available instance is directly
+  /// allocated when true.
+  /// \return Status object indicating success or failure.
+  static Status Create(
+      const bool ignore_resources_and_priority,
+      std::unique_ptr<RateLimiter>* rate_limiter);
+
+  /// Loads the configified model to the RateLimiter
+  /// \param model_name The name of the model to load.
+  /// \param version The version of the model to load.
+  /// \param model_config The configuration of the model to load.
+  /// \return Status object indicating success or failure.
+  Status LoadModel(
+      const std::string& model_name, const int64_t version,
+      const inference::ModelConfig& model_config);
+
+  /// Unloads the configified model from the RateLimiter
+  /// \param model_name The name of the model to load.
+  /// \param version The version of the model to load.
+  /// \return Status object indicating success or failure.
+  Status UnloadModel(const std::string& model_name, const int64_t version);
+
+  /// Enqueues the callback to the configified model. In future, when the
+  /// conditions are met, the callback will be invoked and a pointer to
+  /// allocated RateLimiter::ModelInstance object will be exposed as a
+  /// parameter. The user must ensure RateLimiter::ModelInstance::Release
+  /// gets called on the instance once the inference request is complete
+  /// so that the instance and its resources are returned to the available
+  /// pool. Also, note the callback should be a light-weight call and
+  /// must not itself invoke the inference execution but just be used
+  /// as a signal to proceed with the execution.
+  /// \param OnSchedule The callback function to be called when scheduling.
+  /// \param model_name The name of the model.
+  /// \param version The version of the model.
+  /// \param instance_index The index to a configific instance of the model.
+  /// The default value is -1 which means that an instance with highest
+  /// priority will be selected for the execution.
+  /// \return Status object indicating success or failure.
+  Status EnqueueModelRequest(
+      const StandardScheduleFunc& OnSchedule, const std::string& model_name,
```

Comment:
Using model name, version, instance index... we have adstractions for TritonModel and TritonModelInstance that we could probably be using directly. This is ok for now but we should revisit when integrating this into scheduler to see how we can use those abstractions.

Reply:
That is right. I have implemented it like this for simplicity.
In future it will be more like passing raw pointers to TritonModel/TritonModelInstance to add and remove model instances. We can use the API to attain the name, version and instance index. The TritonModelInstance pointer and additional information(RateLimiterConfig) will be wrapped within RateLimiter::ModelInstance class. This way we would not have to disturb the backend API.